src/util/Histogram.cc

```diff
@@ -0,0 +1,229 @@
+    _checkEntries();
+}
+
+void Histogram::_checkEntries() {
+    auto now = CLOCK::now();
+    auto originalSize = _entries.size();
+    while (true && !_entries.empty()) {
+        Entry& head = _entries.front();
+        auto age = now - head.stamp;
+        if (age.count() > _maxAge.count() || _entries.size() > _maxSize) {
+            auto val = head.val;
+            _total -= val;
+            _changeCountsBy(val, -1);
+            _entries.pop();
+
+        } else {
+            break;
+        }
+    }
+    if (_entries.size() == 0 && originalSize != 0) {
+        // clear values.
+        for (auto& bkt : _buckets) {
+            bkt.count = 0;
+        }
+        _overMaxCount = 0;
+        _total = 0.0;
+    }
+}
+
+int Histogram::getBucketCount(size_t index) const {
+    std::lock_guard<std::mutex> lock(_mtx);
+    if (index > _buckets.size()) {
+        LOGS(_log, LOG_LVL_ERROR, "Histogram::getBucketCount out of range index=" << index);
+        return 0;
+    }
+    if (index == _buckets.size()) {
+        return _overMaxCount;
+    }
+    return _buckets[index].count;
+}
+
+double Histogram::getBucketMaxVal(size_t index) const {
+    std::lock_guard<std::mutex> lock(_mtx);
+    if (index > _buckets.size()) {
+        string eMsg = string("Histogram::getBucketCount out of range index=") + to_string(index);
+        LOGS(_log, LOG_LVL_ERROR, eMsg);
+        return std::numeric_limits<double>::max();
+    }
+    if (index == _buckets.size()) {
+        return std::numeric_limits<double>::max();
+    }
+    return _buckets[index].getMaxVal();
+}
+
+double Histogram::getTotal() const {
+    std::lock_guard<std::mutex> lock(_mtx);
+    return _total;
+}
+
+double Histogram::getSize() const {
+    std::lock_guard<std::mutex> lock(_mtx);
+    return _entries.size();
+}
+
+void Histogram::_changeCountsBy(double val, int incr) {
+    bool found = false;
+    for (auto& bkt : _buckets) {
+        if (val <= bkt.getMaxVal()) {
+            bkt.count += incr;
+            found = true;
+            break;
+        }
+    }
+    if (not found) {
+        _overMaxCount += incr;
+    }
+}
+
+double Histogram::getAvg() const {
+    std::lock_guard<std::mutex> lock(_mtx);
+    return _getAvg();
+}
+
+double Histogram::_getAvg() const {
+    double totalCount = _entries.size();
+    double avg = (totalCount) ? _total / totalCount : 0.0;
+    return avg;
+}
+
+void Histogram::setMaxSize(size_t maxSize) {
+    std::lock_guard<std::mutex> lock(_mtx);
+    _maxSize = maxSize;
+    _checkEntries();
+}
+
+void Histogram::setMaxAge(std::chrono::milliseconds maxAge) {
+    std::lock_guard<std::mutex> lock(_mtx);
+    _maxAge = maxAge;
+    _checkEntries();
+}
+
+std::string Histogram::getString(std::string const& note) {
+    std::lock_guard<std::mutex> lock(_mtx);
+    return _getString(note);
+}
+
+/// _mtx must be locked before calling this function.
+///
+string Histogram::_getString(std::string const& note) {
+    stringstream os;
```

Comment:
I don't like that for a few reasons.

1. You care about what mutex is locked, and the lock guard could easily have the wrong mutex. There needs to be comment with the lock_guard saying what mutex should be locked, or include the mutex in the name of the lock_guard parameter.
2. If someone does pass the wrong mutex in the lock_guard, it's a placebo. They see that and think there ok because there's a lock_guard there.
3. A lock_guard wont work in a unique_lock is needed.
4. There are ways of doing this that actually check that the correct mutex is locked and don't care about lock_guard or unique_lock. They require some effort, but if the  correct mutex is not locked by this thread, it will always trigger. It requires a specially written mutex, but it's compatible with everything we do.

Reply:
For number 1, you may name the parameter after the name of the corresponding mutex to be locked by the lock guard:
```
class Histogram {
public:
  std::string _getString(std::lock_guard<std::mutex> const& lockOnMtx1,
                         std::string const& note);
};
```
This won't guarantee anything (beyond a requirement to lock some mutex). However, seeing this signature of the method in the class's contract (the header file) is hopefully better than having a hidden comment in the CC file like:
```
/// _mtx must be locked before calling this function.
///
string Histogram::_getString(std::string const& note) {
...
```
For number 2, it's true. I see no way to prevent this kind of mistake. My solution is not the panacea for all possible problems that might exist in the multithreaded code. This is just a step towards a better-structured code that helps a developer to avoid some traps (of forgetting to lock a mutex). Passing the lock as a parameter of a method would at least force the developer to think about the synchronization requirements of the called method. It's much easier to overlook the hidden comment in the CC file. Besides, the contemporary IDEs (VS Code, etc.) would happily pop up info on the method from the header file.

For 3, yes, I'm aware of this. That's why I never claimed that this is an ideal solution. The solution was mainly designed for implementing simple synchronized methods like the one you have here.

For 4, I myself would be very interested to learn about these techniques. Those would be useful for cases that go beyond solving the problem of implementing trivially synchronized methods.

Reply:
This post https://stackoverflow.com/questions/21892934/how-to-assert-if-a-stdmutex-is-locked . The answer where they define a new mutex that stores the thread id of the locking thread and has a function to verify that the current thread. If you want I could set that up.

Reply:
Looking at the code, there's a util::Mutex class that does this, but it seems to go further than necessary by keeping a list of locked mutexes. I don't quite understand the purpose of the list of locked mutex. It makes locking much more expensive, and even with the list, it would still be very difficult to figure out what went wrong. `assert(_utilMtx.lockedByCaller());` would be as effective, show where the error occured, and be much faster.  Writing our own macros would probably be worth it as well, or do something like `_utilMtx::lockedByCaller(ERR_LOC)`, that way error messages could be logged or throw.